docs/internal/DistributedArchitectureGuide.md

```diff
@@ -188,15 +188,80 @@ Some concepts are applicable to both cluster and project scopes, e.g. [persisten
+
+Translog files are automatically truncated when they are no longer needed, specifically after all their operations have been persisted by Lucene commits on disk.
+Lucene commits are initiated by flushes (e.g., with the index [Flush API]).
+
+Flushes may also be automatically initiated by Elasticsearch, e.g., if the translog exceeds a configurable size (`INDEX_TRANSLOG_FLUSH_THRESHOLD_SIZE_SETTING`) or age (`INDEX_TRANSLOG_FLUSH_THRESHOLD_AGE_SETTING`),which ultimately truncates the translog as well.
+
+#### Acknowledging writes
+
+A bulk request will repeateadly call ultimately the Engine methods such as `index()` or `delete()` which adds operations to the Translog.
+Finally, the AfterWrite action of the `TransportWriteAction` will call `indexShard.syncAfterWrite()` which will put the last written transloc Location of the bulk request into a `AsyncIOProcessor` that is responsible for gradually fsync'ing the Translog and notifying any waiters.
+Ultimately the bulk request is notified that the translog has fsync'ed passed the requested location, and can continue to acknowledge the bulk request.
+
+#### Translog internals
+
+[History retention]:https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-history-retention.html
+
+Each translog is a sequence of files, each identified by a translog generation ID, each containing a sequence of operations, with the last file open for writes.
+The last file has a part which has been fsync'ed to disk, and a part which has been written but not necessarily fsync'ed yet to disk.
+Each operation is identified by a sequence number (`seqno`), which is monotonically increased by the engine's ingestion functionality.
+A `Checkpoint` file is also maintained, that contains, among other information, the current translog generation ID, and its last fsync'ed operation and location, the minimum translog generation ID, and the minimum and maximum sequence number of operations the sequence of translog generations include.
+When the translog rolls over, e.g., upon the translog file exceeding a configurable size, a new file in the sequence is created for writes, and the last one becomes read-only.
+A new commit flushed to the disk will also induce a translog rollover, since the operations in the translog so far will become eligible for truncation.
+
+A few more words on terminology and classes used around the translog Java package.
+A `Location` of an operation is defined by the translog generation file it is contained in, the offset of the operation in that file, and the number of bytes that encode that operation.
+An `Operation` can be a document indexed, a document deletion, or a no-op operation.
+A `Snapshot` iterator can be created to iterate over a range of requested operation sequence numbers read from the translog files.
+A retention lock can be acquired for [History retention] purposes, e.g., for potentially facilitating a replica shard's recovery, which prohibits truncating the translog files.
```

Comment:
I thought the retention lease tracked the earliest sequence number not replicated so that soft-deletes aren't purged past that point. i.e. that it's not about retaining the translog, but rather about preventing old versions of documents around.
This goes to my confusion above about the use of translog beyond local recovery.

Reply:
That's my understanding as well assuming this is `org.elasticsearch.index.seqno.RetentionLease` that we are talking about. Though the comment [here](https://github.com/elastic/elasticsearch/blob/afe503f02fdf87036ba338ee98dd0495abe85ba1/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L2632-L2637) does say "translog files" . So I am not entirely sure. I'd appreciate clarification. Thanks!

Reply:
See also my [reply here](https://github.com/elastic/elasticsearch/pull/126416/files#r2034159842), but you're right that currently the translog is NOT used for peer recovery and NO retention lock is ever acquired.

Note that in older days, this was not true. See [this PR](https://github.com/elastic/elasticsearch/pull/45136/files#diff-60a8d3503d000a2ab6617c09a69f61b000b3b262d68c674a75c2bf6534691d7aR2617) that removes acquiring a retention lock on the translog.

So removing this line from the PR. And adjusted the comment that Yang mentioned.